src/mongo/db/telemetry/telemetry_thread_base.cpp

```diff
@@ -0,0 +1,359 @@
+
+#define MONGO_LOGV2_DEFAULT_COMPONENT ::mongo::logv2::LogComponent::kDefault
+
+namespace mongo {
+
+namespace {
+
+constexpr StringData kParamName = "perconaTelemetry"_sd;
+constexpr StringData kTelemetryPath = "/usr/local/percona/telemetry/psmdb"_sd;
+
+constexpr StringData kFalse = "false"_sd;
+constexpr StringData kTrue = "true"_sd;
+
+
+// We need this flag to filter out updates from server parameter which can arrive before global
+// service context is set. In other words we need to avoid asserts from getGlobalServiceContext()
+bool updatesEnabled = false;
+
+// mutex to serialize external API calls and access to updatesEnabled
+Mutex mutex = MONGO_MAKE_LATCH("TelemetryThread::mutex");
+
+const auto getTelemetryThread =
+    ServiceContext::declareDecoration<std::unique_ptr<TelemetryThreadBase>>();
+
+// start telemetry thread if it is not running
+void startTelemetryThread_inlock(ServiceContext* serviceContext) {
+    auto* telemetryThread = TelemetryThreadBase::get(serviceContext);
+    if (telemetryThread == nullptr || !telemetryThread->running()) {
+        auto telemetryThread = TelemetryThreadBase::create();
+        telemetryThread->go();
+        TelemetryThreadBase::set(serviceContext, std::move(telemetryThread));
+    }
+}
+
+// stop telemetry thread if it is running
+void stopTelemetryThread_inlock(ServiceContext* serviceContext) {
+    auto* telemetryThread = TelemetryThreadBase::get(serviceContext);
+    if (telemetryThread != nullptr) {
+        telemetryThread->shutdown();
+        TelemetryThreadBase::set(serviceContext, {});
+    }
+}
+
+}  // namespace
+
+
+TelemetryThreadBase::TelemetryThreadBase()
+    : BackgroundJob(false /* selfDelete */),
+      _nextScrape(Date_t::now() + Seconds(perconaTelemetryGracePeriod)),
+      _mutex(MONGO_MAKE_LATCH("TelemetryThread::_mutex")) {}
+
+TelemetryThreadBase* TelemetryThreadBase::get(ServiceContext* serviceCtx) {
+    return getTelemetryThread(serviceCtx).get();
+}
+
+void TelemetryThreadBase::set(ServiceContext* serviceCtx,
+                              std::unique_ptr<TelemetryThreadBase> newTelemetryThread) {
+    auto& telemetryThread = getTelemetryThread(serviceCtx);
+    if (telemetryThread) {
+        invariant(
+            !telemetryThread->running(),
+            "Tried to reset the TelemetryThread without shutting down the previous instance.");
+    }
+
+    telemetryThread = std::move(newTelemetryThread);
+}
+
+void TelemetryThreadBase::run() {
+    const ThreadClient tc(name(), getGlobalServiceContext());
+    LOGV2_DEBUG(29121, 1, "starting {name} thread", "name"_attr = name());
+
+    if (auto status = _initParameters(tc->getServiceContext()); !status.isOK()) {
+        LOGV2_ERROR(29133,
+                    "Telemetry thread failed to initialize. Telemetry will be stopped",
+                    "status"_attr = status);
+        _shuttingDown.store(true);
+    }
+
+    while (!_shuttingDown.load()) {
+        if (Date_t::now() >= _nextScrape) {
```

Comment:
As far as I can understand, the `Date_t::now() >= _nextScrape` condition serves two purposes here:
- handles spurious wakeups of the conditional variable
- implements telemetry grace period

We could achieve the same with slightly simper code by transforming `Date_t::now() >= _nextScrape` into a predicate and passing it as the third_argument to `condition_variable::wait_until`. That implies moving the source code block with the call to `_condvar.wait_until` to the very beginning of the while-loop body.

The only concern I see is calling `wait_until` with a absolute timestamp that is already in the past. However, my reading of SO shows we are going to be safe even in that situation: `wait_until` should quickly return.

Reply:
The second purpose is not about grace period only. In case of `mongod` the _nextScrape value is stored in the DB so if for example DB instance was down for a long period of time and then starts The _nextScrape will be in the past on the first iteration of the loop.
I think the solution with predicate is viable but I prefer to keep existing code as more readable.